File: src/engine/rng.ts
import type { DiceSpec } from './types';

const FNV_OFFSET = 2166136261;
const FNV_PRIME = 16777619;

export function normalizeSeed(seed: string): number {
  let hash = FNV_OFFSET;

  for (let index = 0; index < seed.length; index += 1) {
    hash ^= seed.charCodeAt(index);
    hash = Math.imul(hash, FNV_PRIME) >>> 0;
  }

  return hash === 0 ? 1 : hash;
}

export function nextRngState(state: number): number {
  return (Math.imul(state, 1664525) + 1013904223) >>> 0;
}

export function rollDie(state: number, sides: number): { value: number; state: number } {
  const nextState = nextRngState(state);
  const value = (nextState % sides) + 1;

  return { value, state: nextState };
}

export function rollDice(
  state: number,
  specs: DiceSpec[]
): { values: number[]; total: number; state: number } {
  const values: number[] = [];
  let total = 0;
  let workingState = state;

  for (const spec of specs) {
    for (let index = 0; index < spec.count; index += 1) {
      const roll = rollDie(workingState, spec.sides);
      workingState = roll.state;
      values.push(roll.value);
      total += roll.value;
    }
  }

  return {
    values,
    total,
    state: workingState
  };
}

File: backend/engine/utils/rng.py
from __future__ import annotations

from backend.engine.models.state import DiceSpec

FNV_OFFSET = 2_166_136_261
FNV_PRIME = 16_777_619
UINT32_MASK = 0xFFFFFFFF


def _imul(left: int, right: int) -> int:
    """Match JavaScript's 32-bit integer multiply semantics."""
    return (left * right) & UINT32_MASK


def normalize_seed(seed: str) -> int:
    hash_value = FNV_OFFSET

    for character in seed:
        hash_value ^= ord(character)
        hash_value = _imul(hash_value, FNV_PRIME)

    return hash_value or 1


def next_rng_state(state: int) -> int:
    return (_imul(state, 1_664_525) + 1_013_904_223) & UINT32_MASK


def roll_die(state: int, sides: int) -> tuple[int, int]:
    next_state = next_rng_state(state)
    value = (next_state % sides) + 1
    return value, next_state


def roll_dice(state: int, specs: list[DiceSpec]) -> tuple[list[int], int, int]:
    values: list[int] = []
    total = 0
    working_state = state

    for spec in specs:
        for _ in range(spec.count):
            value, working_state = roll_die(working_state, spec.sides)
            values.append(value)
            total += value

    return values, total, working_state
